### PR TITLE
[WIP] Improve setup implementation

### PR DESCRIPTION
**Pull request by hrsh7th**
### Overview

1. the source must be registered by the `source author`
    - It can be registered after setup.
    - The source will be detected on each keypress.

2. the source must be configured by the `user`.
    - So the user should install the source and add it to the configuration like the below.

```viml
" The source author must register the source automatically.
Plug 'some-source-author/compe-awesome-source'

" The user must configure source manually to enable it.
let g:compe = {}
let g:compe.source = {}
let g:compe.source.awesome_source = v:true
```

### Breaking change

1. `compe#source#vim_bridge#register` is not work now.
- Use `compe#register_source(name, source)` instead.

2. `require"compe":register_lua_source` is not work now.
- Use `require"compe":register_source(name, source)` instead.

3. The source does not be enabled automatically. To enable it needs explicit configuration.
- It will have to do by the user.

### CODE REVIEW

**Comment by hrsh7th** [comment]
I think it is good change to organize some API's.

So I will refactor som codebase before merge this.

**Comment by hrsh7th** [comment]
I will refactor some utilities move to `lua/compe/utils` directory.
And remove `lua/compe/init.lua`.

File: lua/compe/completion/init.lua [hunk 1]
```diff
@@ -254,6 +242,25 @@ function Completion.should_ignore_display(self)
...
+    end
+  end
+
+  table.sort(sources, function(source1, source2)
+    local meta1 = source1:get_metadata()
+    local meta2 = source2:get_metadata()
+    if meta1.priority ~= meta2.priority then
+      return meta1.priority > meta2.priority
+    end
+  end)
```
Comment (hrsh7th) [inline lua/compe/completion/init.lua:260]: It should cache.

Comment (tami5) [reply lua/compe/completion/init.lua:260]: What you mean?

Comment (hrsh7th) [reply lua/compe/completion/init.lua:260]: The get_sources is called frequently.

Comment (tami5) [reply lua/compe/completion/init.lua:260]: Oh I see

**Pull request closed without merging**